File: src/Simulation/models/consumer.ts
import { Types } from "mongoose";
import { DB } from "../DB-Connector/db-connector";
import { IComponent } from "./node";
import { Weather } from "../weather";
import { assert } from "console";
import DefaultNode from "./defaultnode";
import DataMonitor from "../handlers/DataMonitor";

export class Consumer extends DefaultNode {
  timefn: number[];
  consumption: (temp: number) => number;
  profile: number;
  
  constructor(id: string, timefn: number[], demand?: number, profile?: number) {
    super();
    this.id = id;
    assert(timefn.length == 24);
    this.timefn = timefn;
    this.output = 0;
    this.asset = "consumer";

    if (!profile) {
      // generate a normal distrobution value on size of household, with accomplacing constant that determine the rate of consumption,
      // Larger profile value means it consumes more... multiplicative with weather conditions.
      const r = Math.random();
      const rd = Math.random();
      // size is the number of individuals in a house hold as a real number, it is 2.7% weighted,
      const size = r * 4 - r * 2 + rd * 2 + 4;
      // lamba is the variability of the the consumer to consume, this is much more weighted but tends to be a lower value.
      const lamba = r - r / 2 + rd / 2 + 1;
      // the weights has been predetermined from offical consumption statistics, the lamba is a rough estimate of non consumption including non households
      this.profile = size * 0.027 + lamba * 0.5;
    } else this.profile = profile;

    this.consumption = (temp) => {
      // The tempreture x in celsius, profile p, profile at time stamp tf:
      // p * ((21 - x)^2 + tf) + 1
      // tempreture in our case is in kelvin this 294.15 is 21 celsius
      return (
        this.profile * (0.002 * Math.pow(294.15 - temp, 2)) +
        timefn[new Date().getHours()] +
        1
      );
    };
    this.demand = demand ? demand : 0;
    this.tick = (time: number) => {
      this.demand = this.consumption(Weather.Instance.temp);
      if (this.timeToMonitor < time) {
        this.timeToMonitor = time + 10000;
        DataMonitor.instance.status(this as IComponent);
      }
    };
  }

  async document() {
    try {
      const body = {
        demand: this.demand,
        timefn: this.timefn,
        profile: this.profile,
        name: process.env.NAME,
      };
      await DB.Models.Consumer!.findByIdAndUpdate(this.id, body, {
        upsert: true,
      }).exec();
    } catch (error) {
      console.log(error);
    }
  }

  static generateTimeFn(): number[] {
    const random = Math.random();
    return [
      0.02 * random,
      0.0114 * random,
      0.011 * random,
      0.05 * random,
      0.2 * random,
      0.35 * random,
      0.6 * random,
      0.8 * random,
      0.65 * random,
      0.64 * random,
      0.56 * random,
      0.58 * random,
      0.74 * random,
      0.56 * random,
      0.3 * random,
      0.2 * random,
      0.812 * random,
      0.911 * random,
      0.922 * random,
      0.926 * random,
      0.845 * random,
      0.76 * random,
      0.311 * random,
      0.121 * random,
    ]; // profile for each hour of the day
  }
}

File: services/Simulator_migrated/src/models/consumer.rs
use chrono::{Timelike, Utc};
use mongodb::Database;
use rand::Rng;
use serde::{Deserialize, Serialize};

use crate::{handlers::weather_handler::{ WeatherReport}, db::{ consumer_document::ConsumerDocument}};

use super::node::{Asset, Node};

pub type TimeFn = [f64; 24];
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Consumer {
    pub timefn: TimeFn,
    pub profile: f64,
    pub asset: Asset,
    pub demand: f64,
    pub id: String,
    pub network: String,
}

impl Consumer {
    pub fn new(
        timefn: Option<TimeFn>,
        profile: Option<f64>,
        asset: Option<Asset>,
        id: Option<String>,
        network: String,
    ) -> Self {
        if let Some(p) = profile {
            Self {
                timefn: if let Some(tf) = timefn {
                    tf
                } else {
                    Consumer::generate_time_fn()
                },
                profile: p,
                asset: if let Some(a) = asset {
                    a
                } else {
                    Asset::Customer1
                },
                demand: 0.,
                id: if let Some(i) = id {
                    i
                } else {
                    uuid::Uuid::new_v4().to_string()
                },
                network,
            }
        } else {
            let mut rng = rand::thread_rng();
            let r: f64 = rng.gen();
            let rd: f64 = rng.gen();

            // size is the number of individuals in a house hold as a real number, it is 2.7% weighted,
            let size = r * 4. - r * 2. + rd * 2. + 4.;
            // lamba is the variability of the the consumer to consume, this is much more weighted but tends to be a lower value.
            let lamba = r - r / 2. + rd / 2. + 1.;

            Self {
                timefn: if let Some(tf) = timefn {
                    tf
                } else {
                    Consumer::generate_time_fn()
                },
                profile: size * 0.027 + lamba * 0.5,
                asset: if let Some(a) = asset {
                    a
                } else {
                    Asset::Customer1
                },
                demand: 0.,
                id: if let Some(id) = id {
                    id
                } else {
                    uuid::Uuid::new_v4().to_string()
                },
                network,
            }
        }
    }
    pub fn consumption(&self, temp: f64) -> f64 {
        let now = Utc::now();
        let hour = now.hour();

        self.profile * (0.002 * (294.15 - temp).powf(2.) + self.timefn[hour as usize]) / 3.6 // kw 
    }
    pub async fn document(&self, db: Database) -> Result<mongodb::results::UpdateResult, mongodb::error::Error> {
        ConsumerDocument::update(db, self.id.to_string(), self).await
    }

    pub fn generate_time_fn() -> TimeFn {
        let mut rng = rand::thread_rng();
        let random: f64 = rng.gen::<f64>() * 10.;
        return [
            0.02 * random,
            0.0114 * random,
            0.011 * random,
            0.05 * random,
            0.2 * random,
            0.35 * random,
            0.6 * random,
            0.8 * random,
            0.65 * random,
            0.64 * random,
            0.56 * random,
            0.58 * random,
            0.74 * random,
            0.56 * random,
            0.3 * random,
            0.2 * random,
            0.812 * random,
            0.911 * random,
            0.922 * random,
            0.926 * random,
            0.845 * random,
            0.76 * random,
            0.311 * random,
            0.121 * random,
        ];
    }
}

impl Node<Consumer> for Consumer {
    fn tick(&mut self, _elapsed: f64, weather_report: WeatherReport) {
        self.demand = self.consumption(weather_report.temp);
    }

    fn get_asset(&self) -> Asset {
        self.asset
    }

    fn new(obj: Consumer) -> Self {
        obj
    }
}

#[tokio::test]
async fn test_demand() {
    let cm: &mut Consumer = &mut Node::new(Consumer::new(None, None, None, None, format!("")));
    cm.tick(1., WeatherReport { temp: 26.0, wind_speed: 0. });
    assert!(cm.demand > 0.);
}
